src/state_machine/state.hpp

```diff
@@ -39,108 +40,58 @@ class Main;  // Forward declaration
 
 class State {
  public:
-  State(Logger &log, Main *state_machine);
+  State();
+  static State *getInstance();
+  static void initialise();
 
-  void checkEmergencyStop();
+  virtual void enter(Logger &log) = 0;
+  virtual void exit(Logger &log)  = 0;
 
-  virtual void transitionCheck() = 0;
+  virtual State *checkTransition(Logger &log) = 0;
 
-  Logger &log_;
```

Comment:
I just don't think it's the states' job. Everything that has got to do with I/O should be in main and the states should only *do* stuff if they're explicitly asked. It's not a good idea to start throwing unrelated things into an object just because it's more convenient.

Reply:
That's fair, definitely not a good idea to chuck unrelated things into classes for no good reason. But one could also argue that it is a state's job to report on what it is doing for debugging purposes. And most classes in the code base have their own logger even though some of them use it much less than the states do so I think it would be reasonable to make an exception in this case.
It is unfortunate that logging cannot be handled completely separately without polluting actual useful code, but it seems to me that passing the logger as an argument all the time makes that problem worse rather than better. The ultimate solution might be to have some sort of globally accessible logging so you won't need to pass logger instances around at all but that's a whole different issue.

In any case, I don't mind too much, so whatever you decide to (not) do, I'm on board with it. :)

Reply:
Also, not giving each state the Logger makes it easier because we can initialise all the state instances right away and don't rely on Main to pass the right arguments.

Reply:
Ah, ok, that's a good point :+1: